### PR TITLE
More stream reading improvements

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: quinn-proto/src/connection/streams/recv.rs
Comment (Ralith) [inline quinn-proto/src/connection/streams/recv.rs:335]: Could we yield this once at the end, rather than per call to `next`? I understand that you want to make it hard to skip, but there's no way to enforce that a struct field gets looked at, and it seems strange to check this repeatedly when we can only act on it after we're done due to the borrow.

File: quinn-proto/src/connection/streams/recv.rs
Comment (Ralith) [inline quinn-proto/src/connection/streams/recv.rs:244]: Why do we need this?

File: quinn-proto/src/connection/streams/recv.rs
Comment (Ralith) [inline quinn-proto/src/connection/streams/recv.rs:220]: Are we missing a check for whether the stream is stopped here?

File: quinn-proto/src/connection/streams/recv.rs
Comment (Ralith) [inline quinn-proto/src/connection/streams/recv.rs:357]: This looks redundant to the `UnknownStream` case in `ReadableError`; a stream can never become stopped mid-read. It's also confusingly named.

File: quinn-proto/src/tests/mod.rs
Comment (Ralith) [inline quinn-proto/src/tests/mod.rs:248]: Could we just check the `status` of the prior call, rather than calling again?

File: quinn/src/streams.rs
Comment (Ralith) [inline quinn/src/streams.rs:360]: To prevent loss of specific errors, we must check the status, stash any terminal states locally, and generate a suitable error without calling `inner.read` on the next poll. Similarly below.

**Review by Ralith — CHANGES_REQUESTED** [review]
Great work so far!

File: quinn-proto/src/connection/assembler.rs
Comment (Ralith) [inline quinn-proto/src/connection/assembler.rs:141]: We should get rid of the corresponding read-time code, since this makes it redundant.

**Comment by Ralith** [comment]
Looks like there'll be a bit of a rebase hazard with #1000. Shouldn't be disastrous, though.

**Comment by geieredgar** [comment]
I would be fine with waiting until this PR is merged and then rebasing my changes on top of it.

File: quinn-proto/src/connection/streams/recv.rs
Comment (djc) [reply quinn-proto/src/connection/streams/recv.rs:65]: That's right, this is a bit of a behavior change. Would you like me to split it out of this commit?

File: quinn-proto/src/connection/assembler.rs
Comment (djc) [reply quinn-proto/src/connection/assembler.rs:141]: What code do you think that is, exactly? In my mind the read-time code is more general, since it covers things that are duplicated but still after the current `bytes_read`, so I don't think we can just get rid of it?

File: quinn-proto/src/connection/assembler.rs
Comment (djc) [reply quinn-proto/src/connection/assembler.rs:28]: Yup, that seems better.

File: quinn-proto/src/connection/assembler.rs
Comment (djc) [reply quinn-proto/src/connection/assembler.rs:43]: Fixed.

File: quinn-proto/src/connection/assembler.rs
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:141]: I was thinking of the `if (chunk.offset + chunk.bytes.len() as u64) <= self.bytes_read` clause, within the ordered case. On review, though, you're correct: if `bytes_read` is 0 and we receive two frames starting at 0, they aren't guaranteed to be deduplicated until read, so that code should be retained.

**Comment by djc** [comment]
Scoped this down again to remove the hard stuff (last commit) -- all the other comments should be addressed, so we can just get the simple stuff merged. @geieredgar thanks for being accomodating, I hope the rebase isn't too painful.

**Comment by djc** [comment]
(I will take review comments for the last commit in consideration for the new PR as well, of course.)

**Review by Ralith — APPROVED** [review]

**Pull request merged**
